### PR TITLE
Update reentrant exception to be of specific type

### PR DESCRIPTION
**Pull request by BradBuhrkuhl-ms**
Change the existing InvalidOperationException to specific exceptions for first (StackReentrantSemaphoreNestingViolationException) and subsequent (SemaphoreFaultedException) semaphore violations.

### CODE REVIEW

File: src/Microsoft.VisualStudio.Threading/SemaphoreFaultedException.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.Threading
+{
+    using System;
+
+    /// <summary>
+    /// Exception thrown when Semaphore is in a faulted state.
```
Comment (jdrobison) [inline src/Microsoft.VisualStudio.Threading/SemaphoreFaultedException.cs:9]: "Semaphore" should be `<see cref="AsyncSemaphore"/>`

File: src/Microsoft.VisualStudio.Threading/StackReentrantSemaphoreNestingViolationException.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.Threading
+{
+    using System;
+
+    /// <summary>
+    /// Exception which is thrown by the Reentrant Semaphore when it's reentrancy is violated.
```
Comment (jdrobison) [inline src/Microsoft.VisualStudio.Threading/StackReentrantSemaphoreNestingViolationException.cs:9]: Suggest rewording this to be something like, "Exception which is thrown when the reentrancy contract of a `<see cref="ReentrantSemaphore"/>` created with `<see cref="ReentrancyMode.Stack"/>` is violated."

File: src/Microsoft.VisualStudio.Threading/StackReentrantSemaphoreNestingViolationException.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    /// <summary>
+    /// Exception which is thrown when the reentrancy contract of a <see cref="ReentrantSemaphore"/> created with <see cref="ReentrancyMode.Stack"/> is violated.
+    /// </summary>
+    public class StackReentrantSemaphoreNestingViolationException : InvalidOperationException
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="StackReentrantSemaphoreNestingViolationException"/> class.
+        /// </summary>
+        public StackReentrantSemaphoreNestingViolationException()
+            : base(string.Format(Strings.SemaphoreStackNestingViolated, ReentrantSemaphore.ReentrancyMode.Stack))
```
Comment (AArnott) [inline src/Microsoft.VisualStudio.Threading/StackReentrantSemaphoreNestingViolationException.cs:17]: This is missing `CultureInfo.CurrentCulture` as a first argument.

File: src/Microsoft.VisualStudio.Threading/SemaphoreFaultedException.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.Threading
+{
+    using System;
+
+    /// <summary>
+    /// Exception thrown when a <see cref="AsyncSemaphore"/> is in a faulted state.
```
Comment (AArnott) [inline src/Microsoft.VisualStudio.Threading/SemaphoreFaultedException.cs:9]:
```suggestion
    /// Exception thrown when a <see cref="ReentrantSemaphore"/> is in a faulted state.
```

File: src/Microsoft.VisualStudio.Threading/StackReentrantSemaphoreNestingViolationException.cs [hunk 5]
```diff
@@ -0,0 +1,21 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.Threading
+{
+    using System;
+
+    /// <summary>
+    /// Exception which is thrown when the reentrancy contract of a <see cref="ReentrantSemaphore"/> created with <see cref="ReentrancyMode.Stack"/> is violated.
```
Comment (AArnott) [inline src/Microsoft.VisualStudio.Threading/StackReentrantSemaphoreNestingViolationException.cs:9]:
```suggestion
    /// Exception which is thrown when the reentrancy contract of a <see cref="ReentrantSemaphore"/> created with <see cref="ReentrantSemaphore.ReentrancyMode.Stack"/> is violated.
```

**Review by AArnott — CHANGES_REQUESTED** [review]
Looks good. Just a few fixes, one of which will resolve the build break.

**Review by AArnott — APPROVED** [review]

File: src/Microsoft.VisualStudio.Threading/StackReentrantSemaphoreNestingViolationException.cs [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+namespace Microsoft.VisualStudio.Threading
+{
+    using System;
+    using System.Globalization;
+
+    /// <summary>
+    /// Exception which is thrown when the reentrancy contract of a <see cref="ReentrantSemaphore"/> created with <see cref="ReentrantSemaphore.ReentrancyMode.Stack"/> is violated.
+    /// </summary>
+    public class StackReentrantSemaphoreNestingViolationException : InvalidOperationException
```
Comment (AArnott) [inline src/Microsoft.VisualStudio.Threading/StackReentrantSemaphoreNestingViolationException.cs:12]: I'm thinking a bit more now how this is a whole exception type for a *very* specific error condition. What if we call it `IllegalSemaphoreUsageException` and change its public constructor to take the standard `string, Exception` parameters? I think that will still accomplish your goals, but it might be useful for more scenarios in the future so we don't have to keep introducing new public API as error conditions may be discovered.

Comment (BradBuhrkuhl-ms) [reply src/Microsoft.VisualStudio.Threading/StackReentrantSemaphoreNestingViolationException.cs:12]: That is a bit of a philosophical question, but I tend to error on the side of specific errors for specific cases.  I believe in this case, using a more generic exception would be fine as I don't see another way to violate usage of the semaphore in this execution path.  However, it is essential to keep the first fail (that puts things into the failed mode) separate from the subsequent failures (semaphore already broken/faulted).

Comment (BradBuhrkuhl-ms) [reply src/Microsoft.VisualStudio.Threading/StackReentrantSemaphoreNestingViolationException.cs]: Updated to IllegalSemaphoreUsageException

Comment (BradBuhrkuhl-ms) [reply src/Microsoft.VisualStudio.Threading/StackReentrantSemaphoreNestingViolationException.cs]: It has been a bit since I have done a github PR.  I think it needs a re-review after latest changes.  Thanks @AArnott!

File: src/Microsoft.VisualStudio.Threading/IllegalSemaphoreUsageException.cs [hunk 7]
```diff
@@ -0,0 +1,22 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.Threading
+{
+    using System;
+    using System.Globalization;
+
+    /// <summary>
+    /// Exception which is thrown when the contract of a <see cref="AsyncSemaphore"/> is violated.
```
Comment (AArnott) [inline src/Microsoft.VisualStudio.Threading/IllegalSemaphoreUsageException.cs:10]:
```suggestion
    /// Exception which is thrown when the contract of a <see cref="ReentrantSemaphore"/> is violated.
```

**Review by AArnott — APPROVED** [review]
